### PR TITLE
Bug 2045566 - Add Firefox Labs APIs

### PR DESCRIPTION
**Pull request by freshstrangemusic**
### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/releases.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](https://github.com/mozilla/application-services/blob/main/CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due diligence applied in selecting them.

### CODE REVIEW

**Comment by freshstrangemusic** [comment]
This is missing a few more tests and a changelog entry, but its ~ done

File: components/nimbus/src/stateful/firefox_labs.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+pub const FIREFOX_LABS_CONNECT_URL_KEY: &str = "connect";
+
+#[cfg_attr(test, derive(Debug, PartialEq, Eq))]
+pub struct FirefoxLabsMetadata {
+    pub slug: String,
+    pub enrolled: bool,
+    pub requires_restart: bool,
+    pub title_string_id: String,
+    pub description_string_id: String,
+    pub connect_url: Option<String>,
```
Comment (ohall-m) [inline components/nimbus/src/stateful/firefox_labs.rs]: nit: Android calls this `feedback_url`, but it doesn't really matter, since we have to deserialize it to the view type anyway. Fenix generally tries to keep implementation specific names away, when we can.

File: components/nimbus/src/stateful/firefox_labs.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    pub enrollment_change_events: Vec<EnrollmentChangeEvent>,
+}
+
+#[derive(Eq, PartialEq)]
+#[cfg_attr(test, derive(Debug))]
+pub enum FirefoxLabsEnrollStatus {
+    Enrolled,
+    AlreadyEnrolled,
+    NoExperiment,
+    NotFirefoxLabsOptIn,
```
Comment (ohall-m) [inline components/nimbus/src/stateful/firefox_labs.rs:31]: Thanks for chatting with me about these on Slack! Just leaving what we talked about for future reference, please feel free to correct:

* `Enrolled` -> Do nothing, leave toggle enabled
* `AlreadyEnrolled` -> Do nothing, leave toggle enabled 
* `NoExperiment` -> "This Lab is no longer available". Can only happen if there is a background update while the user has the screen open.
* `NotFirefoxLabsOptIn` -> Generic error message, lab remove from screen.
* `Error` -> Generic error message.

I also asked on [Figma again](https://www.figma.com/design/BVqEYexibOA2betNVUMrfl?node-id=18010-588149&m=dev#1739354039) if we want any specific messaging in these situations.

**Review by ohall-m — COMMENTED** [review]
Thanks for working on this!! The API surface looks perfect, thank you!

Just leaving a note that I've began manually testing locally with this patch. My main goal is to test using `nimbus-cli` and pull `getAvailableFirefoxLabs`. I also hope to test connecting `enrollInFirefoxLab` and `unenrollFromFirefoxLab`.

LGTM, mostly just want to test and understand a few more things.

**Review by ohall-m — COMMENTED** [review]
r+ from an Android/Kotlin perspective. [Patch still needs a qualified Nimbus reviewer.]

I manually tested on Android using nimbus-cli and successfully was able to `getAvailableFirefoxLabs` and enroll/unenroll from a cli Labs that set `nimbus-validation`'s `settings-punctuation`.

I didn't test edge cases, but established the general flow works as expected.

File: components/nimbus/src/stateful/firefox_labs.rs (hunk 2 above)
Comment (freshstrangemusic) [reply components/nimbus/src/stateful/firefox_labs.rs:31]: We should also remove the entry in the case of NoExperiment

File: components/nimbus/src/stateful/firefox_labs.rs (hunk 1 above)
Comment (freshstrangemusic) [reply components/nimbus/src/stateful/firefox_labs.rs]: Yep this can be feedback_url, no problem.

Comment (freshstrangemusic) [reply components/nimbus/src/stateful/firefox_labs.rs]: Done

File: components/nimbus/src/schema.rs [hunk 3]
```diff
@@ -23,6 +25,7 @@ pub struct EnrolledExperiment {
     pub user_facing_name: String,
     pub user_facing_description: String,
     pub branch_slug: String,
+    pub is_rollout: bool,
```
Comment (ohall-m) [inline components/nimbus/src/schema.rs:28]: I'm seeing some [EnrolledExperiments](https://searchfox.org/firefox-main/search?q=EnrolledExperiment%28&path=&case=false&regexp=false) complaining in Fenix testing as a heads up for a companion patch or default value!

Comment (freshstrangemusic) [reply components/nimbus/src/schema.rs:28]: Filed https://bugzilla.mozilla.org/show_bug.cgi?id=2048051

**Review by jeddai — APPROVED** [review]
